magento/magento2
# Fix interception for virtual types in non global area

Fixes #17825

### Description
When a VirtualType defined in di.xml located in non global area (e.g. frontend, adminhtml subfolder etc) the plugins for the Base class of such Virtual Type will not work.

This happens because at the moment Object Manager is configured, only Global area di.xml files are loaded:
*Magento/Framework/App/ObjectManager/Environment/Developer.php* Line 76:
```php
        $objectManager->configure(
            $objectManager
                ->get(\Magento\Framework\App\ObjectManager\ConfigLoader::class)
                ->load(Area::AREA_GLOBAL)
        );
        $objectManager->get(\Magento\Framework\Config\ScopeInterface::class)
            ->setCurrentScope('global');
        $diConfig->setInterceptionConfig(
            $objectManager->get(\Magento\Framework\Interception\Config\Config::class)
        );
```
The *Magento\Framework\Interception\Config\Config* Class in constructor will load configs from all scopes and generates *_intercepted* config array. However to support VirtualTypes it uses call:
```php
$this->_omConfig->getOriginalInstanceType($type)
```
Which relies on current ObjectManager Configuration. While at the moment we have only *Area::AREA_GLOBAL* loaded in ObjectManager.
As a result *Interception\Config* does not see any VirtualTypes defined in other Areas and marks classes as not intercepted.
```php
protected function _inheritInterception($type, $force = false)
    {
        $type = ltrim($type, '\\');
        if (!isset($this->_intercepted[$type]) || $force) {
            $realType = $this->_omConfig->getOriginalInstanceType($type);
            if ($type !== $realType) {
                if ($this->_inheritInterception($realType)) {
                    $this->_intercepted[$type] = true;
                    return true;
                }
            } else {
                $parts = explode('\\', $type);
                if (!in_array(end($parts), $this->_serviceClassTypes) && $this->_relations->has($type)) {
                    $relations = $this->_relations->getParents($type);
                    foreach ($relations as $relation) {
                        if ($relation && $this->_inheritInterception($relation)) {
                            $this->_intercepted[$type] = true;
                            return true;
                        }
                    }
                }
            }
            $this->_intercepted[$type] = false;
        }
        return $this->_intercepted[$type];
    }
```

### Fix

The idea is to add *extend()* function to *Magento\Framework\Interception\Config\Config* to be able to regenerate *_intercepted* array for the definitions loaded in the Application class.

This PR is not complete since it only calls *extend* in *Magento\Framework\App\Http* class.
Please feel free to use it as a reference or as a suggestion.

### Fixed Issues
1. magento/magento2[#17825](https://github.com/magento/magento2/issues/17825): Plugins not called for Magento\Catalog\Block\Product\ListProduct in search page in developer mode

### Manual testing scenarios
1. Define a VirtualType in any module 'frontend/di.xml'
2. Define a plugin for a base class of the virtualType (for ex. throw exception or log something in beforeToHtml).
3. Use VirtualType somewhere on frontend. (for ex. add as a Block in layout)
4. Open a page in browser which uses a class you defined plugin and VirtualType for.
5. See exception not thrown or log not written because plugin didn`t work.

**Expected result**
5. See exception thrown or log written in the plugin

### Contribution checklist (*)
